File: 240927/마법의 숲 탐색/magical-forest-exploration.c
#include <stdio.h>

#define MAX_R 75
#define MAX_C 74
#define MAX_K 1001

const int WALL = -1;
const int EMPTY = 0;
const int _dx[] = {-1, 0, 1, 0};
const int _dy[] = {0, 1, 0, -1};

int _r;
int _c;
int _sum;
int _stack_idx;
int _low[MAX_K];
int _stack[MAX_K];
int _visited[MAX_K];
int _out[MAX_K][2];
int _map[MAX_R][MAX_C];

int
max(int num1, int num2) {
	return num1 > num2 ? num1 : num2;
}

int
get_low(int idx, int out_x, int out_y)
{
	int i;
	int n_idx;

	_visited[idx] = 1;
	_stack[_stack_idx++] = idx;
	for (i = 0; i < 4; i++) {
		n_idx = _map[out_x + _dx[i]][out_y + _dy[i]];
		if (n_idx == WALL || n_idx == EMPTY || _visited[n_idx]) {
			continue;
		}
		_low[idx] = max(_low[idx], get_low(n_idx, _out[n_idx][0], _out[n_idx][1]));
	}
	return _low[idx];
}

void
insert(int idx, int x, int y, int d) {
	int i;
	int j;

	for (;;) {
		if (_map[x + 1][y - 1] == EMPTY
				&& _map[x + 1][y + 1] == EMPTY
				&& _map[x + 2][y] == EMPTY) {
			x++;
		} else if (_map[x - 1][y - 1] == EMPTY
				&& _map[x][y - 2] == EMPTY
				&& _map[x + 1][y - 1] == EMPTY
				&& _map[x + 1][y - 2] == EMPTY
				&& _map[x + 2][y - 1] == EMPTY) {
			x++;
			y--;
			d = (d + 3) & 3;
		} else if (_map[x - 1][y + 1] == EMPTY
				&& _map[x][y + 2] == EMPTY
				&& _map[x + 1][y + 1] == EMPTY
				&& _map[x + 2][y + 1] == EMPTY
				&& _map[x + 1][y + 2] == EMPTY) {
			x++;
			y++;
			d = (d + 1) & 3;
		} else {
			break;
		}
	}
	if (x < 4) {
		for (i = 3; i < _r + 3; i++) {
			for (j = 2; j <= _c + 1; j++) {
				_map[i][j] = EMPTY;
			}
		}
		return;
	}
	_map[x][y] = idx;
	_map[x - 1][y] = idx;
	_map[x][y + 1] = idx;
	_map[x + 1][y] = idx;
	_map[x][y - 1] = idx;
	_low[idx] = x - 1;
	_out[idx][0] = x + _dx[d];
	_out[idx][1] = y + _dy[d];
	_sum += get_low(idx, _out[idx][0], _out[idx][1]);
	while (_stack_idx) {
		_visited[_stack[--_stack_idx]] = 0;
	}
}

int
main() {
	int k;
	int y;
	int d;
	int i;
	int j;

	scanf("%d %d %d", &_r, &_c, &k);
	for (i = 0; i < _r + 3; i++) {
		_map[i][0] = WALL;
		_map[i][1] = WALL;
		for (j = 2; j <= _c + 1; j++) {
			_map[i][j] = EMPTY;
		}
		_map[i][_c + 2] = WALL;
		_map[i][_c + 3] = WALL;
	}
	for (i = 0; i < _c + 4; i++) {
		_map[_r + 3][i] = WALL;
		_map[_r + 4][i] = WALL;
	}
	_sum = 0;
	for (i = 0; i <= k; i++) {
		_visited[i] = 0;
	}
	_stack_idx = 0;
	for (i = 1; i <= k; i++) {
		scanf("%d %d", &y, &d);
		insert(i, 1, y + 1, d);
	}
	printf("%d", _sum);
}

File: 240927/마법의 숲 탐색/magical-forest-exploration.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.StringTokenizer;

public class Main {
	private static final int WALL = -1;
	private static final int EMPTY = 0;
	private static final int[] dx = {-1, 0, 1, 0};
	private static final int[] dy = {0, 1, 0, -1};

	private static int r;
	private static int c;
	private static int x;
	private static int y;
	private static int d;
	private static int sum;
	private static int[] low;
	private static int[][] exit;
	private static int[][] map;
	private static boolean[] visited;
	private static ArrayDeque<Integer> q;

	private static int getLow(int idx, int exitX, int exitY) {
		int i;
		int nIdx;

		visited[idx] = true;
		q.addLast(idx);
		for (i = 0; i < 4; i++) {
			nIdx = map[exitX + dx[i]][exitY + dy[i]];
			if (nIdx == WALL || nIdx == EMPTY || visited[nIdx]) {
				continue;
			}
			low[idx] = Math.max(low[idx], getLow(nIdx, exit[nIdx][0], exit[nIdx][1]));
		}
		return low[idx];
	}

	private static void insert(int idx) {
		int i;

		for (;;) {
			if (map[x + 1][y - 1] == EMPTY
					&& map[x + 1][y + 1] == EMPTY
					&& map[x + 2][y] == EMPTY) {
				x++;
			} else if (map[x - 1][y - 1] == EMPTY
					&& map[x][y - 2] == EMPTY
					&& map[x + 1][y - 1] == EMPTY
					&& map[x + 1][y - 2] == EMPTY
					&& map[x + 2][y - 1] == EMPTY) {
				x++;
				y--;
				d = (d + 3) & 3;
			} else if (map[x - 1][y + 1] == EMPTY
					&& map[x][y + 2] == EMPTY
					&& map[x + 1][y + 1] == EMPTY
					&& map[x + 2][y + 1] == EMPTY
					&& map[x + 1][y + 2] == EMPTY) {
				x++;
				y++;
				d = (d + 1) & 3;
			} else {
				break;
			}
		}
		if (x < 4) {
			for (i = 3; i < r + 3; i++) {
				System.arraycopy(map[0], 2, map[i], 2, c);
			}
			return;
		}
		map[x][y] = idx;
		map[x - 1][y] = idx;
		map[x][y + 1] = idx;
		map[x + 1][y] = idx;
		map[x][y - 1] = idx;
		low[idx] = x - 1;
		exit[idx] = new int[] {x + dx[d], y + dy[d]};
		sum += getLow(idx, exit[idx][0], exit[idx][1]);
		while (!q.isEmpty()) {
			visited[q.pollFirst()] = false;
		}
	}

	public static void main(String[] args) throws IOException {
		int k;
		int i;
		BufferedReader br;
		StringTokenizer st;

		br = new BufferedReader(new InputStreamReader(System.in));
		st = new StringTokenizer(br.readLine(), " ", false);
		r = Integer.parseInt(st.nextToken());
		c = Integer.parseInt(st.nextToken());
		k = Integer.parseInt(st.nextToken());
		map = new int[r + 5][c + 4];
		for (i = 0; i < r + 3; i++) {
			map[i][0] = WALL;
			map[i][1] = WALL;
			map[i][c + 2] = WALL;
			map[i][c + 3] = WALL;
		}
		for (i = 0; i < c + 4; i++) {
			map[r + 3][i] = WALL;
			map[r + 4][i] = WALL;
		}
		sum = 0;
		low = new int[k + 1];
		exit = new int[k + 1][];
		visited = new boolean[k + 1];
		q = new ArrayDeque<>();
		for (i = 1; i <= k; i++) {
			st = new StringTokenizer(br.readLine(), " ", false);
			x = 1;
			y = Integer.parseInt(st.nextToken()) + 1;
			d = Integer.parseInt(st.nextToken());
			insert(i);
		}
		System.out.print(sum);
	}
}